cpp/src/arrow/flight/transport/grpc/grpc_client.cc

```diff
@@ -0,0 +1,872 @@
+        finished_(false) {}
+
+  void TryCancel() override { rpc_->context.TryCancel(); }
+
+ protected:
+  Status Finish() override {
+    if (finished_) {
+      return server_status_;
+    }
+
+    // Drain the read side, as otherwise gRPC Finish() will hang. We
+    // only call Finish() when the client closes the writer or the
+    // reader finishes, so it's OK to assume the client no longer
+    // wants to read and drain the read side. (If the client wants to
+    // indicate that it is done writing, but not done reading, it
+    // should use DoneWriting.
+    ReadPayload message;
+    while (internal::ReadPayload(stream_.get(), &message)) {
+      // Drain the read side to avoid gRPC hanging in Finish()
+    }
+
+    server_status_ = internal::FromGrpcStatus(stream_->Finish(), &rpc_->context);
+    if (!server_status_.ok()) {
+      server_status_ = Status::FromDetailAndArgs(
+          server_status_.code(), server_status_.detail(), server_status_.message(),
+          ". gRPC client debug context: ", rpc_->context.debug_error_string());
+    }
+    if (!transport_status_.ok()) {
+      if (server_status_.ok()) {
+        server_status_ = transport_status_;
+      } else {
+        server_status_ = Status::FromDetailAndArgs(
+            server_status_.code(), server_status_.detail(), server_status_.message(),
+            ". gRPC client debug context: ", rpc_->context.debug_error_string(),
+            ". Additional context: ", transport_status_.ToString());
+      }
+    }
+    finished_ = true;
+
+    return server_status_;
+  }
+
+  std::shared_ptr<ClientRpc> rpc_;
+  std::shared_ptr<Stream> stream_;
+  std::shared_ptr<MemoryManager> memory_manager_;
+  bool finished_;
+  Status server_status_;
+  // A transport-side error that needs to get combined with the server status
+  Status transport_status_;
+};
+
+/// \brief A ClientDataStream implementation for gRPC that manages a
+///   mutex to protect from concurrent reads/writes, and drains the
+///   read side on finish.
+template <typename Stream, typename ReadPayload>
+class WritableDataStream : public FinishableDataStream<Stream, ReadPayload> {
+ public:
+  using Base = FinishableDataStream<Stream, ReadPayload>;
+  WritableDataStream(std::shared_ptr<ClientRpc> rpc, std::shared_ptr<Stream> stream,
+                     std::shared_ptr<MemoryManager> memory_manager)
+      : Base(std::move(rpc), std::move(stream), std::move(memory_manager)),
+        read_mutex_(),
+        finish_mutex_(),
+        done_writing_(false) {}
+
+  Status WritesDone() override {
+    // This is only used by the writer side of a stream, so it need
+    // not be protected with a lock.
+    if (done_writing_) {
+      return Status::OK();
+    }
+    done_writing_ = true;
+    if (!stream_->WritesDone()) {
+      // Error happened, try to close the stream to get more detailed info
+      return internal::ClientDataStream::Finish(MakeFlightError(
```

Comment:
Hmm... does it actually call `Finish` on the base class without doing any of the derived handling (such as draining reads before finishing the stream)?

Reply:
If I don't qualify it, it complains it can't find the overload. (Hmm, probably I should just rename one of the methods to avoid this confusion.)

Reply:
Ah, it needs to be this->Finish, I forgot that bit.

Reply:
Oh, ok, I thought you wanted to call the base method instead of the derived one.